File: Day01/TS/main.ts
import * as fs from "fs";

const digits = [
  "one",
  "two",
  "three",
  "four",
  "five",
  "six",
  "seven",
  "eight",
  "nine",
];

function is_numeric(c: string): boolean {
  return c >= "0" && c <= "9";
}

function replaceWordNumbers(inputString: string): string[] {
  const wordToNumber: { [key: string]: string } = {
    one: "1",
    two: "2",
    three: "3",
    four: "4",
    five: "5",
    six: "6",
    seven: "7",
    eight: "8",
    nine: "9",
  };

  let test = inputString;

  const indexes = Object.keys(wordToNumber).map((key, i) => {
    const ind = inputString.indexOf(key);
    if (ind != -1) return [ind, i];
    return -1;
  });

  const indexes2 = Object.keys(wordToNumber).map((key, i) => {
    const ind = inputString.lastIndexOf(key);
    if (ind != -1) return [ind, i];
    return -1;
  });

  const min = indexes.filter((i) => i != -1).sort((a, b) => a[0] - b[0])[0];
  const max = indexes2.filter((i) => i != -1).sort((a, b) => b[0] - a[0])[0];

  if (min !== undefined)
    inputString = inputString.replace(
      digits[min[1]],
      wordToNumber[digits[min[1]]]
    );

  if (max !== undefined) {
    const tmp = test.substring(
      test.lastIndexOf(digits[max[1]]) + digits[max[1]].length,
      test.length
    );

    test =
      test.substring(0, test.lastIndexOf(digits[max[1]])) +
      wordToNumber[digits[max[1]]] +
      tmp;
  }
  return [inputString, test];
}

function findInt(str: string): number {
  let res = 0;
  const strs = replaceWordNumbers(str);
  str = strs[0];
  let tmp = strs[1];

  for (let i = 0; i < str.length; i++) {
    if (is_numeric(str[i])) {
      res += parseInt(str[i]) * 10;
      break;
    }
  }

  for (let i = tmp.length - 1; i >= 0; i--) {
    if (is_numeric(tmp[i])) {
      res += parseInt(tmp[i]);
      break;
    }
  }

  return res;
}

const main = async () => {
  const data = await fs.readFileSync(
    "../../resources/Day01/firstInput",
    "utf8"
  );

  const lines = data.split("\n");
  let res = 0;

  lines.forEach((line) => {
    if (line === "") return;
    res += findInt(line);
  });

  console.log(res);
};

main();

File: Day01/C/main.c
/*
** EPITECH PROJECT, 2023
** AdventOfCode2023
** File description:
** main.c
*/

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include "my.h"

static int is_digit_word(char *line)
{
    char *words[9] = {"one", "two",
        "three", "four", "five", "six",
        "seven", "eight", "nine"
    };
    size_t i = 0;
    int j = 0;

    while (i < strlen(line)) {
        if (j == 9) {
            j = 0;
            i++;
            continue;
        }
        if (strncmp(line, words[j], strlen(words[j])) == 0) {
                return j + 1;
            }
        j++;
    }
    return 0;
}

static int first_digit_compare(char *line, int i, int *res)
{
    int tmp = 0;

    if (IS_DIGIT(line[i])) {
        *res += (line[i] - '0') * 10;
        return 0;
    } else {
        tmp = is_digit_word(&line[i]);
        if (tmp) {
            *res += tmp * 10;
            return 0;
        }
    }
    return 1;
}

static int last_digit_compare(char *line, int end, int *res)
{
    int tmp = 0;

    if (IS_DIGIT(line[end])) {
        *res += (line[end] - '0');
        return 0;
    } else {
        tmp = is_digit_word(&line[end]);
        if (tmp) {
            *res += tmp;
            return 0;
        }
    }
    return 1;
}

static int check_line(char *line)
{
    int end = strlen(line) - 1;
    int res = 0;
    int is_first = 1;
    int is_last = 1;

    for (size_t i = 0; i < strlen(line); i++) {
        if (is_first)
            is_first = first_digit_compare(line, i, &res);
        if (is_last)
            is_last = last_digit_compare(line, end, &res);
        end--;
        if (!is_first && !is_last)
            break;
    }
    return res;
}

int main(void)
{
    FILE *file = fopen("./resources/Day01/firstInput", "r");
    char *buf = NULL;
    int s = 0;
    int res = 0;
    size_t size = 1;

    s = getline(&buf, &size, file);
    while (s != -1) {
        res += check_line(buf);
        s = getline(&buf, &size, file);
    }
    free(buf);
    fclose(file);
    printf("%d\n", res);
    return 0;
}
